pkg/datastore/mysql/mysql.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+
+	// mysql driver
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"go.uber.org/zap"
+)
+
+// MySQL client wrapper
+type MySQL struct {
+	client       *sql.DB
+	logger       *zap.Logger
+	usernameFile string
+	passwordFile string
+}
+
+// Option for create MySQL typed instance
+type Option func(*MySQL)
+
+// WithLogger returns logger setup function
+func WithLogger(logger *zap.Logger) Option {
+	return func(m *MySQL) {
+		m.logger = logger
+	}
+}
+
+// WithAuthenticationFile returns auth info setup function
+func WithAuthenticationFile(usernameFile, passwordFile string) Option {
+	return func(m *MySQL) {
+		m.usernameFile = usernameFile
+		m.passwordFile = passwordFile
+	}
+}
+
+// NewMySQL returns new MySQL instance
+func NewMySQL(url, database string, opts ...Option) (*MySQL, error) {
+	m := &MySQL{
+		logger: zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	dataSourceName, err := m.buildDataSourceName(url, database)
+	if err != nil {
+		return nil, err
+	}
+	db, err := sql.Open("mysql", dataSourceName)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Do you think that we should send a Ping to check the connection?
(It will block a while though.)

Reply:
I think it's not necessary since the error will be raised on unable to connect to remote datastore with wrong `dataSourceName` 🤔

Reply:
👍